mockkube/src/main/java/io/strimzi/test/mockkube2/MockKube2.java

```diff
@@ -0,0 +1,160 @@
+import io.strimzi.api.kafka.Crds;
+import io.strimzi.api.kafka.model.Kafka;
+import io.strimzi.api.kafka.model.KafkaConnect;
+import io.strimzi.api.kafka.model.KafkaConnector;
+import io.strimzi.api.kafka.model.KafkaMirrorMaker2;
+import io.strimzi.api.kafka.model.KafkaRebalance;
+import io.strimzi.api.kafka.model.KafkaTopic;
+import io.strimzi.api.kafka.model.StrimziPodSet;
+import io.strimzi.test.TestUtils;
+import io.strimzi.test.mockkube2.controllers.AbstractMockController;
+import io.strimzi.test.mockkube2.controllers.MockDeploymentController;
+import io.strimzi.test.mockkube2.controllers.MockPodController;
+import io.strimzi.test.mockkube2.controllers.MockServiceController;
+import io.strimzi.test.mockkube2.controllers.MockStatefulSetController;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MockKube2 {
+    private final KubernetesClient client;
+    private final List<AbstractMockController> controllers = new ArrayList<>();
+
+    public MockKube2(KubernetesClient client) {
+        this.client = client;
+    }
+
+    private void addController(AbstractMockController controller)   {
+        controllers.add(controller);
+    }
+
+    private void registerCrd(String apiVersion, String kind, Class<? extends KubernetesResource> crdClass, String crdPath)  {
+        KubernetesDeserializer.registerCustomKind(apiVersion, kind, crdClass);
+
+        CustomResourceDefinition kafkaCrd = client
+                .apiextensions().v1()
+                .customResourceDefinitions()
+                .load(crdPath)
+                .get();
+
+        client.apiextensions().v1().customResourceDefinitions().create(kafkaCrd);
+    }
+
+    public void start() {
+        for (AbstractMockController controller : controllers)   {
+            controller.start();
+        }
+    }
+
+    public void stop() {
+        for (AbstractMockController controller : controllers)   {
+            controller.stop();
+        }
+    }
+
+    public static class MockKube2Builder {
+        private final KubernetesClient client;
+        private final MockKube2 mock;
+
+        public MockKube2Builder(KubernetesClient client) {
+            this.client = client;
+            this.mock = new MockKube2(client);
+        }
+
+        public MockKube2Builder withDeploymentController()  {
+            mock.addController(new MockDeploymentController(client));
+            return this;
+        }
+
+        public MockKube2Builder withStatefulSetController()  {
+            mock.addController(new MockStatefulSetController(client));
+            return this;
+        }
+
+        public MockKube2Builder withPodController()  {
+            mock.addController(new MockPodController(client));
+            return this;
+        }
+
+        public MockKube2Builder withServiceController()  {
+            mock.addController(new MockServiceController(client));
+            return this;
+        }
+
+        public MockKube2Builder withKafkaCrd()  {
```

Comment:
For the CRDs I wonder if we shouldn't:

1. Register them all automatically?
2. Provide an API in the `api` module for discovering all the CRDs. Something as simple as a `static` method that returned a list of some record class of group, Kind, CR class, List class? Using that here any appropriately elsewhere would make it easier to add new CRDs.

Reply:
> Register them all automatically?

TBH, it seems unnecessary. We never use more than 2 CRDs really in the tests. Plus this actually creates the CRDs in the mock server, so I thought that maybe it is better to save the time.

> Provide an API in the api module for discovering all the CRDs. Something as simple as a static method that returned a list of some record class of group, Kind, CR class, List class? Using that here any appropriately elsewhere would make it easier to add new CRDs.

We actually have some part of this. We have for example this [array](https://github.com/strimzi/strimzi-kafka-operator/blob/198e1e639a165934d9e8df1bff1dc1986a5e7521/api/src/main/java/io/strimzi/api/kafka/Crds.java#L40) (although currently private). We have also a method for registering all CRDs for deserialisation. However, the problem is that to register CRD for the mock server, we also need the YAMLs of the CRDs. And this does not exists in the `api` module (as it is only generated by a combination of the `api` module and the `crd-generator`). So I cannot use this directly. I would need to pair it somewhere with the YAMLs first. Any ideas how to do it in some elegant way?

Reply:
We could have a new module (`crds`) that:

1. Depended on the `api` module
2. Included the generated CRDs as classpath resources
3. Included the kind of discovery API I'm talking about

I'm not even sure whether it needs to be a separate module; since we generate the CRDs when building the `api` module we could just add them to the `src/main/resources`. Same-jar is attractive because it makes it impossible for the YAMLs to end up being a different version than the classes (e.g. due to messed up poms).

We might (eventually) want some kind of way to check/assert that the classes in the api module are compatible with the CRDs existing in an apiserver. I'm not meaning Kubernetes' idea of CRD schema compatibility here. It would be to prevent people using more recent jars against and API server with older installed CRDs. A simple way of achieving this would be to annotate the CRDs with an integer version which we increment with each release, and also include that integer in the api module, then check that `crd version from api server <= crd version from jar`.

Reply:
Yes, that would be fine @scholzj. Thanks.

Reply:
Raised #6767